examples/architecture/JSONCtrProperties.wyt

```diff
@@ -0,0 +1,81 @@
+import wyvern.collections.list
+import wyvern.archast
+import wyvern.ast
+import wyvern.option
+
+type JSONCtrProperties
+	val host : String
+	val prt : Int
+	metadata new
+		def checkPortCompatibility(l : list.List[archast.PortDecl]) : Int
+			val length = l.length()
+			type PortDecl = archast.PortDecl
+			var p1 : PortDecl = archast.NonePort
+			var p2 : PortDecl = archast.NonePort
+			if (l.length() == 2)	
+					val op1 : option.Option[PortDecl] = l.get(0)
+					val op2 : option.Option[PortDecl] = l.get(1)
+					if (op1.isDefined && op2.isDefined)
+							p1 = op1.getOrElse(() => p1)
+							p1 = op2.getOrElse(() => p1)
+						else 
+							-1
+					var requires : String = ""
+					var provides : String = ""
+					if (p1.getRequires() == "")
+							if (p1.getProvides() == "")
+									-1
+								else
+									provides = p1.getProvides()
+						else
+							requires = p1.getRequires()
+					if (requires == "")	
+							requires = p2.getRequires()
+						else	
+							provides = p2.getProvides()
+								
+					if (requires == provides)
+							length
+						else
+							-1
+					length
+				else
+					-1
```

Comment:
The way you have done the compatibility check is pretty convoluted. The way I would have done this is as follows:

1. Have 2 boolean variables: one for flagging if there is a requires port and the other for flagging if there is provides port. Initialize both of them to false.

2. Check if op1 is a requires or provides port and set the appropriate boolean variable.

3. Check if op2 is a requires or provides port and set the appropriate boolean variable.

4. Return the conjunction of the two boolean variables.

Reply:
I think I might be able to simplify the nested if blocks a little bit, but I think to do it like how you describe above we aren't able to check that the ports are requiring/providing the same interface. Also I had figured since op1 and op2 are options wrapping the PortDecl object we can't get to the PortDecl and therefore the requires/provides of it without unwrapping and providing a default (the NonePort in archast). Is there a workaround for this?

Reply:
Ah, yes! You are right! But this additional check can also be integrated into the steps I listed above. You can have variables for holding the interface for each port. Then, in steps 2 and 3, you can set these variables. Finally, in step 4, you can add an additional check to see whether the two interfaces are the same.